src/image/loading_displaying.js

```diff
@@ -158,6 +159,363 @@ p5.prototype.loadImage = function(path, successCallback, failureCallback) {
+ *   background(hue, 40, 60);
+
+ *   // create a circle that moves diagonally
+ *   circle(
+ *     100 * sin(frameCount / 10) + width / 2,
+ *     100 * sin(frameCount / 10) + height / 2,
+ *     10
+ *   );
+ * }
+
+ * // you can put it in the mousePressed function,
+ * // or keyPressed for example
+ * function mousePressed() {
+ *   // this will download the first two seconds of my animation!
+ *   saveGif('mySketch', 2, { units: 'seconds', delay: 0 });
+ * }
+ * </code>
+ * </div>
+ *
+ * @alt
+ * animation of a circle moving smoothly diagonally
+ */
+p5.prototype.saveGif = async function(
+  fileName,
+  duration,
+  { delay = 0, units = 'seconds' }
+) {
+  // validate parameters to throw friendly error
+  p5._validateParameters('saveGif', arguments);
+
+  // get the project's framerate
+  let _frameRate = this._targetFrameRate;
+  // if it is undefined or some non useful value, assume it's 60
+  if (_frameRate === Infinity || _frameRate === undefined || _frameRate === 0) {
+    _frameRate = 60;
+  }
+
+  // calculate frame delay based on frameRate
+
+  // this delay has nothing to do with the
+  // delay in options, but rather is the delay
+  // we have to specify to the gif encoder between frames.
+  let gifFrameDelay = 1 / _frameRate * 1000;
+
+  // constrain it to be always greater than 20,
+  // otherwise it won't work in some browsers and systems
+  // reference: https://stackoverflow.com/questions/64473278/gif-frame-duration-seems-slower-than-expected
+  gifFrameDelay = gifFrameDelay < 20 ? 20 : gifFrameDelay;
+
+  // check the mode we are in and how many frames
+  // that duration translates to
+  const nFrames = units === 'seconds' ? duration * _frameRate : duration;
+  const nFramesDelay = units === 'seconds' ? delay * _frameRate : delay;
+
+  // initialize variables for the frames processing
+  let count = nFramesDelay;
+
+  // we start on the frame set by the delay argument
+  //   frameCount = nFramesDelay;
+
+  const lastPixelDensity = this._pixelDensity;
+  this.pixelDensity(1);
+
+  // We first take every frame that we are going to use for the animation
+  let frames = [];
+
+  if (document.getElementById('progressBar') !== null)
+    document.getElementById('progressBar').remove();
+
+  let p = this.createP('');
+  p.id('progressBar');
+
+  p.style('font-size', '16px');
+  p.style('font-family', 'Montserrat');
+  p.style('background-color', '#ffffffa0');
+  p.style('padding', '8px');
+  p.style('border-radius', '10px');
+  p.position(0, 0);
+
+  let pixels;
+  let gl;
+  if (this.drawingContext instanceof WebGLRenderingContext) {
+    // if we have a WEBGL context, initialize the pixels array
+    // and the gl context to use them inside the loop
+    gl = document.getElementById('defaultCanvas0').getContext('webgl');
+    pixels = new Uint8Array(gl.drawingBufferWidth * gl.drawingBufferHeight * 4);
+  }
+
+  // stop the loop since we are going to manually redraw
+  this.noLoop();
```

Comment:
cannot we let the loop run as it is and meanwhile we extract the pixel colors ?
using this maybe - https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Pixel_manipulation_with_canvas

Reply:
We are doing exactly what you say! We are enabling the loop back just before extracting the pixel colors and start encoding.

What we cannot do is loop while recording the frames, because we need all the frames in order, one by one, and we extract the frame information from the current Canvas context.

Reply:
I think it would be possible to use the existing `loop()`, but I believe that would require augmenting `draw()` to keep track of the export frame count, and to record a frame at the end of each export frame. That said, I think I personally prefer this method where we manually call `redraw()`, as it keeps `draw()` simple and also it means you can read the gif saving function more easily without having to jump around the codebase to see the data flow.